[Flaky] Fix "should preempt a workload from LQ with higher recent usage"

IrvingMg opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug
/kind flake

#### What this PR does / why we need it:

Fix flaky test "Scheduler when Using AdmissionFairSharing at Cohort level should preempt a workload from LQ with higher recent usage"

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #6227 

#### Special notes for your reviewer:

Giving some extra time to allow localqueue usage to be updated correctly, seem to fix the flakiness.

After applying the fix, the test was run for ~1 hour with 475 attempts, and no flakes occurred.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

IrvingMg commented:
/cc @mimowo

mimowo commented on test/integration/singlecluster/scheduler/fairsharing/fair_sharing_test.go:
```diff
@@ -499,6 +499,14 @@ var _ = ginkgo.Describe("Scheduler", func() {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(lqB), lqB)).Should(gomega.Succeed())
+					usageA := lqA.Status.FairSharing.AdmissionFairSharingStatus.ConsumedResources[corev1.ResourceCPU]
+					usageB := lqB.Status.FairSharing.AdmissionFairSharingStatus.ConsumedResources[corev1.ResourceCPU]
+					g.Expect(usageA.MilliValue()).To(gomega.BeNumerically(">", usageB.MilliValue()))
```
cc @PBundyra

PBundyra replied:
> So, does it mean that we may observe usage(LQa)> 0, usage(LQb)>0, yet usage(LQa) < usage(LQb) for a while? If so, do you know how this is possible?

Yes, this could be the reason. This could be possible because entry penalty for a workload A is higher than for a workload B

IrvingMg replied:
Yeah... I need to check. My assumption was that by adding a new assertion, we could give a bit of extra time for the LocalQueue usages to fully update. But you're right,  with the > 0 assertions, we should already have it fully covered as we create only one workload per localqueue.

Still, it seems to be a sync issue, because with this extra assertion, the tests have been running without flakes so far. I’ll confirm the root cause.

IrvingMg replied:
I've opened a PR with increased logging #6328

mimowo replied:
Actually, thinking about the intention of this specific test - we should choose to preempt the workload which has a higher usage, it is perfectly reasonable to await for the usage to stabilize.

I would even go further, and instead of awaiting for >0, wait until >12 (for workload A). This would simplify the overall conditions in the test (I don't think we even need to test >0 for workload B, but we could.

@IrvingMg @PBundyra wdyt?

IrvingMg commented:
I've reproduced the error with some additional logs, specifically "Calculated LocalQueueFSUsage for workload". You can see that in scheduling cycle 3, lq-a's usage is lower than lq-b's. Then, in subsequent cycles, usageA is updated and remains higher.

Below are the logs for cycles 3 and 4:

```none
STEP: Checking that the workload from lq-A is preempted despite having bigger priority @ 07/31/25 15:31:47.022
  2025-07-31T15:31:47.02217+02:00	LEVEL(-2)	core/workload_controller.go:157	Reconcile Workload	{"controller": "workload_controller", "namespace": "core-6ntpz", "name": "workload-6gpj9", "reconcileID": "7d959af3-23a8-4bf8-b89a-9147f14b9a56"}
  2025-07-31T15:31:47.022389+02:00	LEVEL(-5)	controller/controller.go:355	Reconcile done, requeueing after 1s	{"controller": "localqueue_controller", "namespace": "core-6ntpz", "name": "lq-b", "reconcileID": "8040489b-5308-4c97-9f1d-69de887096ed"}
  2025-07-31T15:31:47.02246+02:00	LEVEL(-2)	localqueue-reconciler	core/localqueue_controller.go:228	Queue update event	{"localQueue": {"name":"lq-b","namespace":"core-6ntpz"}}
  2025-07-31T15:31:47.022485+02:00	LEVEL(-5)	controller/controller.go:339	Reconciling	{"controller": "localqueue_controller", "namespace": "core-6ntpz", "name": "lq-c", "reconcileID": "ed132de4-03b2-4fe2-b374-af30dceb1dc3"}
  2025-07-31T15:31:47.022523+02:00	LEVEL(-2)	core/localqueue_controller.go:152	Reconcile LocalQueue	{"controller": "localqueue_controller", "namespace": "core-6ntpz", "name": "lq-c", "reconcileID": "ed132de4-03b2-4fe2-b374-af30dceb1dc3"}
  2025-07-31T15:31:47.022444+02:00	LEVEL(-5)	controller/controller.go:368	Reconcile successful	{"controller": "workload_controller", "namespace": "core-6ntpz", "name": "workload-6gpj9", "reconcileID": "7d959af3-23a8-4bf8-b89a-9147f14b9a56"}
  2025-07-31T15:31:47.022462+02:00	LEVEL(-5)	core/workload_controller.go:1012	Workload cluster queue update event	{"localQueue": {"name":"lq-b","namespace":"core-6ntpz"}}
  2025-07-31T15:31:47.022222+02:00	LEVEL(-5)	scheduler	cache/snapshot.go:220	Calculated LocalQueueFSUsage for workload	{"schedulingCycle": 3, "workload": {"name":"workload-67g2h","namespace":"core-6ntpz"}, "queue": "lq-a", "usage": 1.339}
  2025-07-31T15:31:47.023296+02:00	LEVEL(-5)	scheduler	cache/snapshot.go:220	Calculated LocalQueueFSUsage for workload	{"schedulingCycle": 3, "workload": {"name":"workload-hbgcc","namespace":"core-6ntpz"}, "queue": "lq-b", "usage": 1.566}
  2025-07-31T15:31:47.023392+02:00	LEVEL(-5)	scheduler	preemption/preemption.go:560	Comparing workloads by LocalQueue fair sharing usage	{"schedulingCycle": 3, "workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "clusterQueue": {"name":"cq2"}, "workloadA": {"name":"workload-hbgcc","namespace":"core-6ntpz"}, "queueA": "lq-b", "usageA": 1.566, "workloadB": {"name":"workload-67g2h","namespace":"core-6ntpz"}, "queueB": "lq-a", "usageB": 1.339}
  2025-07-31T15:31:47.023427+02:00	LEVEL(-5)	scheduler	preemption/preemption.go:387	Simulating fair preemption	{"schedulingCycle": 3, "workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "clusterQueue": {"name":"cq2"}, "candidates": [{"name":"workload-hbgcc","namespace":"core-6ntpz"},{"name":"workload-67g2h","namespace":"core-6ntpz"}], "resourcesRequiringPreemption": [{"Flavor":"default","Resource":"cpu"}], "preemptingWorkload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}}
  2025-07-31T15:31:47.023543+02:00	LEVEL(-5)	scheduler	preemption/preemption.go:560	Comparing workloads by LocalQueue fair sharing usage	{"schedulingCycle": 3, "workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "clusterQueue": {"name":"cq2"}, "workloadA": {"name":"workload-hbgcc","namespace":"core-6ntpz"}, "queueA": "lq-b", "usageA": 1.566, "workloadB": {"name":"workload-67g2h","namespace":"core-6ntpz"}, "queueB": "lq-a", "usageB": 1.339}
  2025-07-31T15:31:47.023573+02:00	LEVEL(-5)	scheduler	preemption/preemption.go:387	Simulating fair preemption	{"schedulingCycle": 3, "workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "clusterQueue": {"name":"cq2"}, "candidates": [{"name":"workload-hbgcc","namespace":"core-6ntpz"},{"name":"workload-67g2h","namespace":"core-6ntpz"}], "resourcesRequiringPreemption": [{"Flavor":"default","Resource":"cpu"}], "preemptingWorkload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}}
  2025-07-31T15:31:47.023614+02:00	LEVEL(-5)	scheduler	scheduler/fair_sharing_iterator.go:81	Computing DominantResourceShare for tournament	{"schedulingCycle": 3, "rootCohort": {"name":"all"}}
  2025-07-31T15:31:47.023642+02:00	LEVEL(-3)	scheduler	scheduler/fair_sharing_iterator.go:84	Running tournament to decide next workload to consider in scheduling cycle	{"schedulingCycle": 3, "rootCohort": {"name":"all"}}
  2025-07-31T15:31:47.023667+02:00	LEVEL(-3)	scheduler	scheduler/fair_sharing_iterator.go:92	Determined tournament winner	{"schedulingCycle": 3, "rootCohort": {"name":"all"}, "cohort": {"name":"all"}, "clusterQueue": {"name":"cq2"}, "winningWorkload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}}
  2025-07-31T15:31:47.023692+02:00	LEVEL(-5)	scheduler	scheduler/fair_sharing_iterator.go:226	DominantResourceShare values used during tournament	{"schedulingCycle": 3, "rootCohort": {"name":"all"}, "cohort": {"name":"all"}, "clusterQueue": {"name":"cq2"}, "winningWorkload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "drsValues": ["{parentCohort: all, workload core-6ntpz/workload-6gpj9, drs: 0}"]}
  2025-07-31T15:31:47.023721+02:00	LEVEL(-2)	scheduler	scheduler/scheduler.go:259	Attempting to schedule workload	{"schedulingCycle": 3, "workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "clusterQueue": {"name":"cq2"}, "parentCohort": {"name":"all"}, "rootCohort": {"name":"all"}}
  2025-07-31T15:31:47.025968+02:00	LEVEL(-5)	localqueue-reconciler	core/localqueue_controller.go:310	Updated LocalQueue fair sharing status	{"namespace": "core-6ntpz", "name": "lq-c", "consumedResources": {}}
  2025-07-31T15:31:47.028441+02:00	LEVEL(-5)	admission	admission/http.go:117	received request	{"webhookGroup": "kueue.x-k8s.io", "webhookKind": "Workload", "Workload": {"name":"workload-hbgcc","namespace":"core-6ntpz"}, "namespace": "core-6ntpz", "name": "workload-hbgcc", "resource": {"group":"kueue.x-k8s.io","version":"v1beta1","resource":"workloads"}, "user": "admin", "requestID": "d02473ab-130f-425a-aef4-7a1d17c641b9"}
  2025-07-31T15:31:47.028723+02:00	LEVEL(-5)	admission.workload-webhook	webhooks/workload_webhook.go:96	Validating update	{"webhookGroup": "kueue.x-k8s.io", "webhookKind": "Workload", "Workload": {"name":"workload-hbgcc","namespace":"core-6ntpz"}, "namespace": "core-6ntpz", "name": "workload-hbgcc", "resource": {"group":"kueue.x-k8s.io","version":"v1beta1","resource":"workloads"}, "user": "admin", "requestID": "d02473ab-130f-425a-aef4-7a1d17c641b9"}
  2025-07-31T15:31:47.028829+02:00	LEVEL(-5)	admission	admission/http.go:163	wrote response	{"webhookGroup": "kueue.x-k8s.io", "webhookKind": "Workload", "code": 200, "reason": "", "message": "", "requestID": "d02473ab-130f-425a-aef4-7a1d17c641b9", "allowed": true}
  2025-07-31T15:31:47.029021+02:00	LEVEL(-5)	controller/controller.go:355	Reconcile done, requeueing after 1s	{"controller": "localqueue_controller", "namespace": "core-6ntpz", "name": "lq-c", "reconcileID": "ed132de4-03b2-4fe2-b374-af30dceb1dc3"}
  2025-07-31T15:31:47.029065+02:00	LEVEL(-5)	core/workload_controller.go:1012	Workload cluster queue update event	{"localQueue": {"name":"lq-c","namespace":"core-6ntpz"}}
  2025-07-31T15:31:47.029068+02:00	LEVEL(-5)	controller/controller.go:339	Reconciling	{"controller": "localqueue_controller", "namespace": "core-6ntpz", "name": "lq-a", "reconcileID": "3c9d97fe-6577-408f-8ba1-fc6e01022795"}
  2025-07-31T15:31:47.029134+02:00	LEVEL(-2)	core/localqueue_controller.go:152	Reconcile LocalQueue	{"controller": "localqueue_controller", "namespace": "core-6ntpz", "name": "lq-a", "reconcileID": "3c9d97fe-6577-408f-8ba1-fc6e01022795"}
  2025-07-31T15:31:47.029082+02:00	LEVEL(-2)	localqueue-reconciler	core/localqueue_controller.go:228	Queue update event	{"localQueue": {"name":"lq-c","namespace":"core-6ntpz"}}
  2025-07-31T15:31:47.030411+02:00	LEVEL(-3)	scheduler	preemption/preemption.go:175	Preempted	{"schedulingCycle": 3, "workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "clusterQueue": {"name":"cq2"}, "parentCohort": {"name":"all"}, "rootCohort": {"name":"all"}, "targetWorkload": {"name":"workload-hbgcc","namespace":"core-6ntpz"}, "preemptingWorkload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "reason": "InCohortFairSharing", "message": "Preempted to accommodate a workload (UID: e4981306-1eff-4712-8412-4edbf7dddf7e, JobUID: UNKNOWN) due to Fair Sharing within the cohort", "targetClusterQueue": {"name":"cq1"}}
  2025-07-31T15:31:47.030494+02:00	DEBUG	events	recorder/recorder.go:104	Preempted to accommodate a workload (UID: e4981306-1eff-4712-8412-4edbf7dddf7e, JobUID: UNKNOWN) due to Fair Sharing within the cohort	{"type": "Normal", "object": {"kind":"Workload","namespace":"core-6ntpz","name":"workload-hbgcc","uid":"ecd6c158-b0cd-4952-a51e-b21937b2971f","apiVersion":"kueue.x-k8s.io/v1beta1","resourceVersion":"257"}, "reason": "Preempted"}
  2025-07-31T15:31:47.030488+02:00	LEVEL(-3)	scheduler	scheduler/logging.go:43	Workload evaluated for admission	{"schedulingCycle": 3, "workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "clusterQueue": {"name":"cq2"}, "status": "", "reason": "couldn't assign flavors to pod set main: insufficient unused quota for cpu in flavor default, 10 more needed. Pending the preemption of 1 workload(s)", "nominatedAssignment": {"PodSets":[{"Name":"main","Flavors":{"cpu":{"Name":"default","Mode":1,"TriedFlavorIdx":-1}},"Status":{},"Requests":{"cpu":"10"},"Count":1,"TopologyAssignment":null,"DelayedTopologyRequest":null}],"Borrowing":0,"LastState":{"LastTriedFlavorIdx":[{"cpu":-1}],"ClusterQueueGeneration":1},"Usage":{"Quota":{"{\"Flavor\":\"default\",\"Resource\":\"cpu\"}":10000},"TAS":null}}, "preempted": [{"name":"workload-hbgcc","namespace":"core-6ntpz"}]}
  2025-07-31T15:31:47.030562+02:00	LEVEL(-2)	scheduler	scheduler/scheduler.go:768	Workload re-queued	{"schedulingCycle": 3, "workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "clusterQueue": {"name":"cq2"}, "queue": {"name":"cq2","namespace":"core-6ntpz"}, "requeueReason": "PendingPreemption", "added": true, "status": ""}
  2025-07-31T15:31:47.030589+02:00	LEVEL(-2)	workload-reconciler	core/workload_controller.go:737	Workload update event	{"workload": {"name":"workload-hbgcc","namespace":"core-6ntpz"}, "queue": "lq-b", "status": "admitted", "clusterQueue": "cq1"}
  2025-07-31T15:31:47.03069+02:00	LEVEL(-5)	controller/controller.go:339	Reconciling	{"controller": "workload_controller", "namespace": "core-6ntpz", "name": "workload-hbgcc", "reconcileID": "77f775aa-db28-46fa-a88f-9b5ef0b1f17c"}
  2025-07-31T15:31:47.030722+02:00	LEVEL(-2)	core/workload_controller.go:157	Reconcile Workload	{"controller": "workload_controller", "namespace": "core-6ntpz", "name": "workload-hbgcc", "reconcileID": "77f775aa-db28-46fa-a88f-9b5ef0b1f17c"}
  2025-07-31T15:31:47.030763+02:00	LEVEL(-5)	controller/controller.go:368	Reconcile successful	{"controller": "workload_controller", "namespace": "core-6ntpz", "name": "workload-hbgcc", "reconcileID": "77f775aa-db28-46fa-a88f-9b5ef0b1f17c"}
  2025-07-31T15:31:47.032276+02:00	LEVEL(-5)	localqueue-reconciler	core/localqueue_controller.go:310	Updated LocalQueue fair sharing status	{"namespace": "core-6ntpz", "name": "lq-a", "consumedResources": {"cpu":"2627m"}}
  2025-07-31T15:31:47.033388+02:00	LEVEL(-5)	admission	admission/http.go:117	received request	{"webhookGroup": "kueue.x-k8s.io", "webhookKind": "Workload", "Workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "namespace": "core-6ntpz", "name": "workload-6gpj9", "resource": {"group":"kueue.x-k8s.io","version":"v1beta1","resource":"workloads"}, "user": "admin", "requestID": "82f7787d-e0c7-447f-9f5f-41a6bd34660f"}
  2025-07-31T15:31:47.033568+02:00	LEVEL(-5)	admission.workload-webhook	webhooks/workload_webhook.go:96	Validating update	{"webhookGroup": "kueue.x-k8s.io", "webhookKind": "Workload", "Workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "namespace": "core-6ntpz", "name": "workload-6gpj9", "resource": {"group":"kueue.x-k8s.io","version":"v1beta1","resource":"workloads"}, "user": "admin", "requestID": "82f7787d-e0c7-447f-9f5f-41a6bd34660f"}
  2025-07-31T15:31:47.033608+02:00	LEVEL(-5)	admission	admission/http.go:163	wrote response	{"webhookGroup": "kueue.x-k8s.io", "webhookKind": "Workload", "code": 200, "reason": "", "message": "", "requestID": "82f7787d-e0c7-447f-9f5f-41a6bd34660f", "allowed": true}
  2025-07-31T15:31:47.034943+02:00	DEBUG	events	recorder/recorder.go:104	couldn't assign flavors to pod set main: insufficient unused quota for cpu in flavor default, 10 more needed. Pending the preemption of 1 workload(s)	{"type": "Warning", "object": {"kind":"Workload","namespace":"core-6ntpz","name":"workload-6gpj9","uid":"e4981306-1eff-4712-8412-4edbf7dddf7e","apiVersion":"kueue.x-k8s.io/v1beta1","resourceVersion":"266"}, "reason": "Pending"}
  2025-07-31T15:31:47.034986+02:00	LEVEL(-2)	workload-reconciler	core/workload_controller.go:737	Workload update event	{"workload": {"name":"workload-6gpj9","namespace":"core-6ntpz"}, "queue": "cq2", "status": "pending"}
  2025-07-31T15:31:47.035157+02:00	LEVEL(-3)	scheduler	queue/manager.go:631	Obtained ClusterQueue heads	{"schedulingCycle": 4, "count": 1}
  2025-07-31T15:31:47.035177+02:00	LEVEL(-2)	localqueue-reconciler	core/localqueue_controller.go:228	Queue update event	{"localQueue": {"name":"lq-a","namespace":"core-6ntpz"}}
  2025-07-31T15:31:47.035178+02:00	LEVEL(-5)	core/workload_controller.go:1012	Workload cluster queue update event	{"localQueue": {"name":"lq-a","namespace":"core-6ntpz"}}
  2025-07-31T15:31:47.035203+02:00	LEVEL(-5)	scheduler	cache/snapshot.go:220	Calculated LocalQueueFSUsage for workload	{"schedulingCycle": 4, "workload": {"name":"workload-67g2h","namespace":"core-6ntpz"}, "queue": "lq-a", "usage": 2.6270000000000002}
  2025-07-31T15:31:47.03522+02:00	LEVEL(-5)	controller/controller.go:355	Reconcile done, requeueing after 1s	{"controller": "localqueue_controller", "namespace": "core-6ntpz", "name": "lq-a", "reconcileID": "3c9d97fe-6577-408f-8ba1-fc6e01022795"}
  2025-07-31T15:31:47.035235+02:00	LEVEL(-5)	controller/controller.go:339	Reconciling	{"controller": "workload_controller", "namespace": "core-6ntpz", "name": "workload-6gpj9", "reconcileID": "3887e90d-841f-47c5-bcf9-28ff44b827c8"}
  2025-07-31T15:31:47.035327+02:00	LEVEL(-2)	core/workload_controller.go:157	Reconcile Workload	{"controller": "workload_controller", "namespace": "core-6ntpz", "name": "workload-6gpj9", "reconcileID": "3887e90d-841f-47c5-bcf9-28ff44b827c8"}
  2025-07-31T15:31:47.035264+02:00	LEVEL(-5)	scheduler	cache/snapshot.go:220	Calculated LocalQueueFSUsage for workload	{"schedulingCycle": 4, "workload": {"name":"workload-hbgcc","namespace":"core-6ntpz"}, "queue": "lq-b", "usage": 1.566}
```

I'm also attaching the full logs.
[out.log](https://github.com/user-attachments/files/21531755/out.log)

PBundyra commented:
I've analyzed the logs and came to the interesting findings. There was a race in processing reconcile requests between `lq-a` and `lq-b`. At some point because of difference of 1 microsecond, lq-b was processed before lq-a.

Here are the logs:

```2025-07-31T15:31:46.783528+02:00  LEVEL(-5) controller/controller.go:355  Reconcile done, requeueing after 216.476ms  {"controller": "localqueue_controller", "namespace": "core-6ntpz", "name": "lq-a", "reconcileID": "ac6e58bc-1bec-44c9-adf5-56cc595c52e4"}```

The request will be ready at `2025-07-31T15:31:47.000004+02:00`

```2025-07-31T15:31:46.786479+02:00  LEVEL(-5) controller/controller.go:355  Reconcile done, requeueing after 213.524ms  {"controller": "localqueue_controller", "namespace": "core-6ntpz", "name": "lq-b", "reconcileID": "754ed359-a24f-488f-9764-f9a24599f500"}```

The request will be ready at `2025-07-31T15:31:47.000003+02:00`

That's why the test sometimes flaked. Because in the first round lq-b sometimes won the race. After that the difference in requested CPU would be too much.

To conclude how we should proceed now:

Following @mimowo 's advice: https://github.com/kubernetes-sigs/kueue/pull/6304#discussion_r2247051775 let's change the assert in the tests so it checks if the lq-a usage is > 12

At the same time let's shorten the half-life time to 1s so it doesn't take dozens of second for the usage to grow above 12

Thanks @IrvingMg for providing logs

PBundyra commented:
/hold cancel

mimowo commented on test/integration/singlecluster/scheduler/fairsharing/suite_test.go:
```diff
@@ -63,48 +73,55 @@ var _ = ginkgo.BeforeSuite(func() {
-		scheduler.WithFairSharing(fairSharing), scheduler.WithAdmissionFairSharing(admissionFairSharing))
-	err = sched.Start(ctx)
-	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+func restartManagerWithDefaultConfig() {
```
This function is used only ones, let's inline

IrvingMg replied:
Done.

mimowo commented on test/integration/singlecluster/scheduler/fairsharing/suite_test.go:
```diff
@@ -63,48 +73,55 @@ var _ = ginkgo.BeforeSuite(func() {
+	}
+}
+
+func restartManagerWithConfig(admissionFairSharing *config.AdmissionFairSharing) {
```
Actually, I don't like the approach of restarting manager, this could introduce more unrelated flakes. 

It also makes tests take llonger for no reason.

Can we instead start the manger with the desired config?

IrvingMg replied:
Currently, all the tests in the suite use the same config values, since the AFS config is loaded in the manager. I'm not sure if it's possible to update the config without restarting the manager.

I think the easiest way to start the manager with the desired config would be to create a separate test suite, but that would require duplicating more code.

mimowo replied:
> I think the easiest way to start the manager with the desired config would be to create a separate test suite, but that would require duplicating more code.

Duplicating beforeEach and AfterEach is fine. These are very repeatable

mimowo replied:
We can have a new When("Half Lifetime is short") or somehing like this. 

It is better to duplicate and avoid restarting manager, for sure.

mimowo commented:
Awesome work!
/lgtm
/approve
/cherry-pick release-0.13
/cherry-pick release-0.12

mimowo commented:
@IrvingMg the cherrypicks will likely need to be manual due to conflicts. Please proceed. I think it is worth it to align the structure of the tests as we will need to maintain 0.12 for another 2 months. So it will make future cherrypicks much easier.

IrvingMg commented:
> Thank you for looking into that.It is a signal the tests were already too convoluted at it is good to restructure

Yeah... The issue was that the test "should promote a workload from LQ with lower recent usage" was implicitly relying on cq2 to be present in order to borrow resources, even though it doesn't explicitly reference cq2. When I separated it from the test "should preempt a workload from LQ with higher recent usage", I removed cq2, which caused the first test to fail.

A simple fix would be to create a second CohortQueue, but since both tests were originally under When("Using AdmissionFairSharing at Cohort level"), I think it's better to keep them together. So I moved them to the new Describe block with the shortened half-life time.

Maybe to better structure the tests, I should move the ones using AdmissionFairSharing into the second Describe block? WDYT?

mimowo commented:
/unhold
/lgtm
Thank you 👍 Merging as this will already fix a common flake
